integrations/redis.go

```diff
@@ -0,0 +1,68 @@
+package integrations
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-redis/redis/v7"
+
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+// RedisSpanHook is a redis.Hook for wrapping Redis commands and pipelines
+// in Client Spans and metrics.
+type RedisSpanHook struct {
+	ClientName string
+}
+
+var _ redis.Hook = RedisSpanHook{}
+
+// BeforeProcess starts a client Span before processing a Redis command and
+// starts a timer to record how long the command took.
+func (h RedisSpanHook) BeforeProcess(ctx context.Context, cmd redis.Cmder) (context.Context, error) {
+	return h.startChildSpan(ctx, cmd.Name()), nil
+}
+
+// AfterProcess ends the client Span started by BeforeProcess, publishes the
+// time the Redis command took to complete, and a metric indicating whether the
+// command was a "success" or "fail"
+func (h RedisSpanHook) AfterProcess(ctx context.Context, cmd redis.Cmder) error {
+	return h.endChildSpan(ctx, cmd.Err())
+}
+
+// BeforeProcessPipeline starts a client span before processing a Redis pipeline
+// and starts a timer to record how long the pipeline took.
+func (h RedisSpanHook) BeforeProcessPipeline(ctx context.Context, cmds []redis.Cmder) (context.Context, error) {
+	return h.startChildSpan(ctx, "pipeline"), nil
+}
+
+// AfterProcessPipeline ends the client span started by BeforeProcessPipeline,
+// publishes the time the Redis pipeline took to complete, and a metric
+// indicating whether the pipeline was a "success" or "fail"
+func (h RedisSpanHook) AfterProcessPipeline(ctx context.Context, cmds []redis.Cmder) error {
+	var err error
+	for _, cmd := range cmds {
+		err = cmd.Err()
+		if err != nil {
+			break
+		}
```

Comment:
:no_bell: Currently we just return the first error, I wonder if we should combine all the errors instead. Konrad did [this in secrets package](https://github.com/reddit/baseplate.go/blob/ada824d60917800f49abc204a1d3a3733fb4549d/secrets/secrets.go#L167-L177), I also wrote [a package for combining errors before](https://godoc.org/github.com/fishy/errbatch).